x-pack/solutions/observability/plugins/apm/public/components/shared/trace_waterfall/use_trace_waterfall.ts

```diff
@@ -89,18 +99,34 @@ export function createColorLookupMap(legends: IWaterfallLegend[]): Map<string, s
   return new Map(legends.map((legend) => [`${legend.type}:${legend.value}`, legend.color]));
 }
 
-export function getTraceParentChildrenMap(traceItems: TraceItem[]) {
-  const traceMap = traceItems.reduce<Record<string, TraceItem[]>>((acc, item) => {
-    if (!item.parentId) {
-      acc.root = [item];
+export function getTraceParentChildrenMap(traceItems: TraceItem[], filteredTrace: boolean) {
+  if (traceItems.length === 0) {
+    return {};
+  }
+
+  const traceMap: Record<string, TraceItem[]> = {};
+
+  traceItems.reduce<Record<string, TraceItem[]>>((acc, item) => {
+    if (item.parentId) {
+      (acc[item.parentId] ??= []).push(item);
     } else {
-      if (!acc[item.parentId]) {
-        acc[item.parentId] = [];
-      }
-      acc[item.parentId].push(item);
+      (acc.root ??= [])[0] = item;
     }
     return acc;
-  }, {});
+  }, traceMap);
```

Comment:
This looks wrong, you are changing the `traceMap` per reference. Can you simplify it by keeping the same old logic?
```

const traceMap = traceItems.reduce<Record<string, TraceItem[]>>((acc, item) => {
...
}. {});
```

Reply:
`??=` will only create a new array when there is nothing present (lazy initialisation). Also in the changed case, if we have multiple roots, the `root` array is initialised only once, with the root item being replaced (yielding the last one as the valid root item like with the previous behaviour).